crates/fj-viewer/src/graphics/model.rs

```diff
@@ -0,0 +1,226 @@
+    pub name: String,
+    pub diffuse_texture: super::texture::Texture,
+    pub bind_group: wgpu::BindGroup,
+}
+
+#[derive(Debug)]
+pub struct Mesh {
+    pub name: String,
+    pub vertex_buffer: wgpu::Buffer,
+    pub index_buffer: wgpu::Buffer,
+    pub num_elements: u32,
+    pub material: usize,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum LoadModelError {
+    #[error("Object loading error")]
+    ObjLoad(#[from] LoadError),
+
+    #[error("Load texture error")]
+    Texture(#[from] LoadTextureError),
+}
+
+pub fn load_model(
+    file_name: &str,
+    device: &wgpu::Device,
+    queue: &wgpu::Queue,
+    layout: &wgpu::BindGroupLayout,
+) -> Result<Model, LoadModelError> {
+    let assets = Assets::get_instance();
+
+    let (models, obj_materials) = tobj::load_obj_buf(
+        &mut assets.get_asset(file_name),
+        &tobj::LoadOptions {
+            triangulate: true,
+            single_index: true,
+            ..Default::default()
+        },
+        |p| {
+            tobj::load_mtl_buf(
+                &mut assets.get_asset(
+                    p.file_name()
+                        .unwrap()
+                        .to_str()
+                        .expect("OsStr could not be converted to a str"),
+                ),
+            )
+        },
+    )?;
+
+    let mut materials = Vec::new();
+    for m in obj_materials? {
+        let texture_data: &[u8] = assets.get_asset(m.diffuse_texture.as_str());
+
+        let diffuse_texture = texture::Texture::from_bytes(
+            device,
+            queue,
+            texture_data,
+            file_name,
+        )?;
+
+        let bind_group = device.create_bind_group(&wgpu::BindGroupDescriptor {
+            layout,
+            entries: &[
+                wgpu::BindGroupEntry {
+                    binding: 0,
+                    resource: wgpu::BindingResource::TextureView(
+                        &diffuse_texture.view,
+                    ),
+                },
+                wgpu::BindGroupEntry {
+                    binding: 1,
+                    resource: wgpu::BindingResource::Sampler(
+                        &diffuse_texture.sampler,
+                    ),
+                },
+            ],
+            label: None,
+        });
+
+        materials.push(Material {
+            name: m.name,
+            diffuse_texture,
+            bind_group,
+        })
+    }
+
+    let meshes = models
+        .into_iter()
+        .map(|m| {
+            let vertices = (0..m.mesh.positions.len() / 3)
+                .map(|i| ModelVertex {
+                    position: [
+                        m.mesh.positions[i * 3],
+                        m.mesh.positions[i * 3 + 1],
+                        m.mesh.positions[i * 3 + 2],
+                    ],
+                    tex_coords: [
+                        m.mesh.texcoords[i * 2],
+                        1.0 - m.mesh.texcoords[i * 2 + 1],
+                    ],
+                })
```

Comment:
Those magic numbers here (specifically, the `3` used for vertex position and the `2` used for texture coordinates) aren't great. It's easy enough to figure out why they are here (number of vertices in a triangle and number of texture coordinates in 2 dimensions, respectively), but it gave me pause and looks a bit iffy.

It might make sense to extract them into appropriately named constants, and extract the `i * 3`/`i * 2` into variables. Although I don't have ideas on what to name those.

Anyway, I'm nitpicking!